File: 78-subsets/subsets.ts
function subsets(nums: number[]): number[][] {
    let n = 2 ** nums.length;
    let ans = [];
    for (let i = 0; i < n; i++) {
        let curArr = [];
        for (let j = 0; j < nums.length; j++) {
            let bitMask = 1 << j;
            let bit = ((bitMask & i) === 0) ? 0 : 1;
            if (bit === 1) curArr.push(nums[j]);
        }
        ans.push(curArr);
    }
    return ans;
};

File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> curList = new ArrayList<>();
        generate(0 , ans , curList , nums);
        return ans;
    }

    private void generate(int index, List<List<Integer>> ans, List<Integer> curList, int[] nums) {
        if(index == nums.length) {
            ans.add(new ArrayList<>(curList));
            return;
        };
        generate(index + 1 , ans ,curList, nums);
        curList.add(nums[index]);
        generate(index + 1 , ans , curList , nums);
        curList.remove(curList.size() - 1);

    }

}